compiler/rustc_trait_selection/src/traits/fulfill.rs

```diff
@@ -767,6 +770,31 @@ impl<'a, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'tcx> {
                         }
                     }
                 }
+                ty::PredicateKind::Clause(ty::ClauseKind::UnstableFeature(symbol)) => {
+                    // Iterate through all goals in param_env to find the one that has the same symbol.
+                    for pred in obligation.param_env.caller_bounds().iter() {
+                        if let ty::ClauseKind::UnstableFeature(sym) = pred.kind().skip_binder() {
+                            if sym == symbol {
+                                return ProcessResult::Changed(Default::default());
+                            }
+                        }
+                    }
+
+                    // During codegen we must assume that all feature bounds hold as we may be
+                    // monomorphizing a body from an upstream crate which had an unstable feature
+                    // enabled that we do not.
+                    //
+                    // Note: we don't consider a feature to be enabled
+                    // if we are in std/core even if there is a corresponding `feature` attribute on the crate.
```

Comment:
future work: we likely want an `#[rustc_allow_internal_unstable(foo)]` attribute which only adds `UnstableFeature` clauses to the param_env, but not `predicates_of`.